src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1546,16 +1546,31 @@
        new_value, old_value, message, file, line, containment_path, containing_class, corrective_change
      from resource_events"]
    (fn [rows]
-     (reduce (fn [hashes-seen row]
-               (conj hashes-seen
-                     (let [hash-str (hash/resource-event-identity-pkey row)]
-                       (when-not (hashes-seen hash-str)
-                         (jdbc/insert-multi! "resource_events_transform"
-                                             (list (assoc row :event_hash
-                                                          (sutils/munge-hash-for-storage hash-str)))))
-                       hash-str)))
-             #{}
-             rows)))
+     (let [old-cols '(:report_id :certname_id :status :timestamp :resource_type
+                      :resource_title :property :new_value :old_value :message
+                      :file :line :containment_path :containing_class
+                      :corrective_change)
+           new-cols (conj old-cols :event_hash)
+           update-row (apply juxt (conj old-cols
+                                        (comp sutils/munge-hash-for-storage
+                                              hash/resource-event-identity-pkey)))
+           insert->hash (fn [batch]
+                          (jdbc/insert-multi! :resource_events_transform
+                                              new-cols
+                                              batch)
+                          ;; Return hashes for each row directly instead of
+                          ;; reading them back from the DB via the return
+                          ;; value of insert-multi!
+                          (map :event_hash batch))
+           dedupe-and-insert (fn [hashes-seen batch]
+                               (->> batch
+                                    (pmap update-row)
+                                    (filter #(-> % first hashes-seen nil?))
+                                    insert->ids
+                                    (reduce conj hashes-seen)))]
+       (reduce dedupe-and-insert
+               #{}
+               (partition 1000 1000 [] rows)))))
```

Comment:
Using a fixed batch size of 1000 here. For the 5.7 million row dataset, this dropped migration time from about 45 minutes to about 15 minutes. I haven't played with the sensitivity of the batch size extensively, but going to 10,000 rows per batch only yielded another 1.5 minutes in speedup. Larger batches will require more RAM to process through PuppetDB.

Reply:
Think I'd be inclined to stick with 1k if the further increase is that marginal.